### PR TITLE
ci-connector-ops: declare review requirements on source connectors changes

### LINKED ISSUE #20304: ci-connector-ops: require reviews from connector team
## Tell us about the problem you're trying to solve
We want to automatically block a PR from being merged without the approval of the Connector Extensibility /Operations team when the backward compatibility checks are disabled.

## Describe the solution you’d like
* Add a new command to `ci_connector_ops` to check if the current branch disabled backward compatibility tests in `acceptance-test-config.yml`
* Create the following logic in a GitHub action:
```
if script returns 1: 
  if not any([reviewers in connectors_team for reviewers in approving_reviewers])
    fail job
```

### PR DESCRIPTION
**Pull request by alafanechere**
## What
Closes https://github.com/airbytehq/airbyte/issues/20304

This [GitHub action](https://github.com/Automattic/action-required-review) can help us check if a set of teams gave a review on a PR by creating a "review requirements" file.
I made a new command in `ci-connector-ops` to write this file according to the current changes in the branch.

I changed the `connector_ops_ci.yml` workflow to:
* Call the new `write-review-requirements-file` command to write the requirements file. This file structure is defined by the `Automattic/action-required-review@v3` action
* Run the `Automattic/action-required-review@v3` if this file was written. 

When the requirements are not met, a status check as this one will appear:
Screen Shot 2022-12-12 at 13 48 59

## The new following review requirements rules:
* If a PR touches a GA connector: a review from `gl-python` team is required
* If a PR changes the test strictness level in `acceptance-test-config.yml`: a review from `connector-operations `teams is required
* If a PR changes backward compatibility test in `acceptance-test-config.yml`: a review from `connector-operations` or `connector-extensibility` is required.

### `ci_connector_ops` python package
* The new command to write the review requirements file: https://github.com/airbytehq/airbyte/blob/9da488761dae17b709d7057f51bba8a85dd2a0de/tools/ci_connector_ops/ci_connector_ops/sat_config_checks.py#L76
* Implement `get_changed_acceptance_test_config` to track changes on `acceptance-test-config.yml`: 
https://github.com/airbytehq/airbyte/blob/9da488761dae17b709d7057f51bba8a85dd2a0de/tools/ci_connector_ops/ci_connector_ops/utils.py#L55
* "Integration" tests:
https://github.com/airbytehq/airbyte/blob/9da488761dae17b709d7057f51bba8a85dd2a0de/tools/ci_connector_ops/tests/test_sat_config_checks.py#L90

### `connector_ops_ci.yml` GitHub action workflow 
* Add a new `check-review-requirements` job
https://github.com/airbytehq/airbyte/blob/45d990a063a5bc2dddee9306318be3a96fd00655/.github/workflows/connector_ops_ci.yml#L25

### CODE REVIEW

File: tools/ci_connector_ops/setup.py
Comment (erohmensing) [inline tools/ci_connector_ops/setup.py:23]: Even if it has been reviewed by a team member, won't it still return 1 and make the build not green? 

I'm not sure exactly how it would look, but maybe in the github action we can do something like (pseudocode in python-ish that has to be bash)
```
if script returns 1: 
  if not any([reviewers in connectors_team for reviewers in approving_reviewers])
    fail job
```

Comment (alafanechere) [reply tools/ci_connector_ops/setup.py:23]: Yes the outer logic of when to call this command should be built in a GitHub action... 
@pedroslopez was suggesting to at least request for review to our team if the script returns 1. Getting the list of approving reviewers looks a bit trickier. Is is something you've done in GHA already?

Comment (erohmensing) [reply tools/ci_connector_ops/setup.py:23]: You're right, I assumed there'd be some sort of readily available `github.event.pull.request.reviewers` variable, but there doesn't seem to be anything regarding reviewers [here](https://docs.github.com/en/actions/using-workflows/events-that-trigger-workflows). 

Another option that may be easier would be instead of checking the review itself, the action could be the one that requests the team for review if the script returns 1. There seem to be a decent amount of actions available that do that:
* https://github.com/marketplace/actions/auto-request-review seems maintained
  * `Auto-assign the default reviewers if no reviewers are matched to your rules` or `Request review only in certain conditions` is probably the relevant behavior here 
  * https://github.com/marketplace/actions/add-pull-request-reviewer is a bit older but seems simple

**Review by erohmensing — APPROVED** [review]
Testing, testing, 1,2,3...

File: .github/workflows/connector_ops_ci.yml
Comment (evantahler) [inline .github/workflows/connector_ops_ci.yml:31]: I've done this kind of thing in the past:

```
if: ${{ github.repository == 'airbytehq/airbyte' }}
```

File: .github/workflows/connector_ops_ci.yml
Comment (evantahler) [inline .github/workflows/connector_ops_ci.yml:56]: [non-blocking] Do you think that we are /always/ the proper people to take the review?  On one hand, we know the most about the spec and possibly the CDK (+connector-extensibility?), but perhaps the question is connector specific and we will need GL...

I guess we can re-assign the review as needed.

File: .github/workflows/connector_ops_ci.yml
Comment (evantahler) [inline .github/workflows/connector_ops_ci.yml:47]: [not-blocking] The way python stringifies "true" is so strange.  Maybe the output should be "true" or "1"?

File: tools/ci_connector_ops/ci_connector_ops/utils.py
Comment (evantahler) [inline tools/ci_connector_ops/ci_connector_ops/utils.py:64]: cool use of git to see if the file changed!

**Review by evantahler — APPROVED** [review]
This is great!

I have the feeling we'll be iterating on /who/ needs to do the required review, but this is an excellent first step.

File: .github/workflows/connector_ops_ci.yml
Comment (alafanechere) [reply .github/workflows/connector_ops_ci.yml:56]: A cool evolution would be to make the python script return the name of the teams that should review. To keep things simple at the moment, I'll make both `connector-operations` and `connector-extensibility` required reviewers (in an `any-of` manner).

File: .github/workflows/connector_ops_ci.yml
Comment (alafanechere) [reply .github/workflows/connector_ops_ci.yml:47]: 👌 I'll make it return `true` or `false`

File: tools/ci_connector_ops/ci_connector_ops/utils.py
Comment (alafanechere) [reply tools/ci_connector_ops/ci_connector_ops/utils.py:64]: The current limitation is that it checks if the `acceptance-test-config.yml` was modified, not specifically that `disable-for-version` was added. We'd be required a review even if the `disable-for-version` key was added in previous PRs and the change to `acceptance-test-config.yml` are not related to backward compatibility. I'll try to check the diff content itself.

File: .github/workflows/connector_ops_ci.yml
Comment (alafanechere) [reply .github/workflows/connector_ops_ci.yml:31]: 👍 Thanks!

File: tools/ci_connector_ops/ci_connector_ops/sat_config_checks.py
Comment (erohmensing) [inline tools/ci_connector_ops/ci_connector_ops/sat_config_checks.py:62]:
```suggestion
    if test_strictness_level_changes:
```

File: tools/ci_connector_ops/ci_connector_ops/sat_config_checks.py
Comment (erohmensing) [inline tools/ci_connector_ops/ci_connector_ops/sat_config_checks.py:84]: Cool idea to generate this on each run!

File: tools/ci_connector_ops/ci_connector_ops/sat_config_checks.py
Comment (erohmensing) [inline tools/ci_connector_ops/ci_connector_ops/sat_config_checks.py:98]: Doing the `any-of` check here feels a little un-DRY, since you also manually add `any-of` in `find_mandatory_reviewers` -- maybe you could put this logic of adding the any-of into a helper method you call in `find_mandatory_reviewers`?

File: tools/ci_connector_ops/ci_connector_ops/sat_config_checks.py
Comment (erohmensing) [inline tools/ci_connector_ops/ci_connector_ops/sat_config_checks.py:99]:
```suggestion
    print(f"MANDATORY_REVIEWERS=A review is required from these teams: {','.join(teams)}")

```

File: .github/workflows/connector_ops_ci.yml
Comment (erohmensing) [inline .github/workflows/connector_ops_ci.yml:51]: To be clear, generating and outputting this separately is a requirement of the github action, right? We can't generate that into `connector_org_review_requirements.yaml`?

**Review by erohmensing — APPROVED** [review]
This looks like it will be super helpful! Might be worth sharing this action and how you implemented it (Loom?) to other teams who may be unsatisfied with how static codeowners is

File: .github/workflows/connector_ops_ci.yml
Comment (alafanechere) [reply .github/workflows/connector_ops_ci.yml:51]: Exactly, I added the  `MANDATORY_REVIEWERS` to conveniently display a detailed status like `A review is required from teams: A, B, C` instead of the default `Review required` status. But all the essential information is captured in the `REVIEW_REQUIREMENTS_FILE` whose format is defined by the external action.

Comment (erohmensing) [reply .github/workflows/connector_ops_ci.yml:51]: Cool, I figured. Strange that you can't configure different status messages for the different sets of requirements in the action 😄

**Comment by YowanR** [comment]
This makes sense at a high-level to me. Adding @lazebnyi as a reviewer to make sure that GL is plugged into this conversation.

**Review by lazebnyi — APPROVED** [review]

**Review by evantahler — APPROVED** [review]

**Review by evantahler — CHANGES_REQUESTED** [review]
I left a comment and that triggered a failing check!

https://github.com/airbytehq/airbyte/actions/runs/3680412195

```
[Invalid workflow file: .github/workflows/connector_ops_ci.yml#L49](https://github.com/airbytehq/airbyte/actions/runs/3680412195/workflow)
The workflow is not valid. .github/workflows/connector_ops_ci.yml (Line: 49, Col: 13): Unexpected symbol: '""'. Located at position 74 within expression: steps.write-review-requirements-file.outputs.REVIEW_REQUIREMENTS_FILE != ""
```

**Comment by alafanechere** [comment]
@evantahler, there was a glitch in the boolean check to conditionally run the final action. I fixed it in https://github.com/airbytehq/airbyte/pull/20303/commits/e7202ff10dff46b8128f1e82c6f60e8c33a29e31

**Review by evantahler — APPROVED** [review]
👍 assuming I don't get another failing comment :D

**Pull request merged**
